### PR TITLE
Delete all subpages when allow child pages without parent is true

### PR DESCRIPTION
**Pull request by InnaBoitsun**
[AB#72515](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/72515)

### CODE REVIEW

File: Composite.Workflows/Plugins/Elements/ElementProviders/PageElementProvider/DeletePageWorkflow.cs [hunk 1]
```diff
@@ -95,6 +96,10 @@ private void codeActivity1_ExecuteCode(object sender, EventArgs e)
             IPage selectedPage = (IPage)dataEntityToken.Data;
 
             bool hasChildren = PageServices.GetChildren(selectedPage.Id).Any();
```
Comment (neexite) [inline Composite.Workflows/Plugins/Elements/ElementProviders/PageElementProvider/DeletePageWorkflow.cs]: It looks like we need to check hasSubPages (not childred)

File: Composite/Data/Types/PageServices.cs [hunk 2]
```diff
@@ -389,6 +409,22 @@ public static IEnumerable<IPage> GetSubChildren(this IPage parentPage)
                         yield return subPage;
                     }
                 }
+
+                if(GlobalSettingsFacade.AllowChildPagesTranslationWithoutParent)
+                {
+                    foreach (Guid emptyPageId in parentPage.GetEmptyChildren())
+                    {
+                        foreach (IPage childPage in GetChildren(emptyPageId))
```
Comment (neexite) [inline Composite/Data/Types/PageServices.cs]: Looks like in the sample "parent1 -> empty -> empty -> grandchild1" grandchild1 will not be returned.

File: Composite/Data/Types/PageServices.cs [hunk 3]
```diff
@@ -81,6 +82,25 @@ orderby ps.LocalOrdering
 #warning revisit this - we return all versions (by design so far). Any ordering on page versions? - check history for original intent
         }
 
+        public static IQueryable<Guid> GetEmptyChildren(this IPage page)
+        {
+            Verify.ArgumentNotNull(page, nameof(page));
+            var parentId = page.Id;
+            var structure = DataFacade.GetData<IPageStructure>();
+            var pagesIds = DataFacade.GetData<IPage>().Select(p => p.Id).ToList();
```
Comment (neexite) [inline Composite/Data/Types/PageServices.cs]: may be too expecive

File: Composite/Data/Types/PageServices.cs [hunk 4]
```diff
@@ -389,7 +390,57 @@ public static IEnumerable<IPage> GetSubChildren(this IPage parentPage)
...
+                        foreach (var page in GetSubChildrenInEmptyParent(emptyPageId, pagesLookup))
+                        {
+                            yield return page;
+                        }
+                    }
+                }
+            }
+        }
+
+        public static IEnumerable<IPage> GetSubChildrenInEmptyParent(Guid emptyParentId, ILookup<Guid, IPage> pages)
```
Comment (neexite) [inline Composite/Data/Types/PageServices.cs]: Can we make it private?

File: Composite/Data/Types/PageServices.cs [hunk 5]
```diff
@@ -389,7 +390,57 @@ public static IEnumerable<IPage> GetSubChildren(this IPage parentPage)
...
+            {
+                foreach (IPage subPage in GetSubChildrenInEmptyParent(emptyPageId, pages))
+                {
+                    yield return subPage;
+                }
+
+            }
+        }
+
+        public static IQueryable<Guid> GetEmptyChildren(Guid parentId, ILookup<Guid, IPage> pages)
```
Comment (neexite) [inline Composite/Data/Types/PageServices.cs]: Can we make it private?

**Review by neexite — APPROVED** [review]

**Comment by burningice2866** [comment]
Its called descendents, not subchildren.

**Pull request merged**